### PR TITLE
proposal: support restoring SQL text from an AST tree

### PR DESCRIPTION
**Pull request by leoppro**
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8533)

### CODE REVIEW

**Comment by leoppro** [comment]
@tiancaiamao @GregoryIan  PTAL

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
+# Proposal: Support to restore SQL text from a AST tree.
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: s/a/an

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+- Last updated:  2018-11-29
+- Discussion at: https://github.com/pingcap/tidb/issues/8532
+
+## Abstract
+
+This proposal aims to support to restore SQL text from any `ast.Node`.
+
+## Background
+
+We know there is a `Text()` function of `ast.Node`,
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: Maybe s/function/method in the Go idiom ?

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+This proposal aims to support to restore SQL text from any `ast.Node`.
+
+## Background
+
+We know there is a `Text()` function of `ast.Node`, 
+the parser calls `SetText()` function during the parsing process, then we can call `Text()` to get original input sql. 
+But the `Text()` function is incomplete, only the `Text()` of root nodes can work well. 
+
+Now some features(eg. Database name and table name mapping) need us to make any `ast.Node` could restore to SQL text.
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: need the `ast.Node` to be restorable to SQL text

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+## Proposal
+
+We know the AST is a type of tree structure of which child node corresponding to SQL text.
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: a kind of tree whose child node is correspond to

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+the parser calls `SetText()` function during the parsing process, then we can call `Text()` to get original input sql. 
+But the `Text()` function is incomplete, only the `Text()` of root nodes can work well. 
+
+Now some features(eg. Database name and table name mapping) need us to make any `ast.Node` could restore to SQL text.
+
+## Proposal
+
+We know the AST is a type of tree structure of which child node corresponding to SQL text. 
+So we can walk through AST tree layer by layer and splice SQL text according to AST node. 
+There is a AST node parsed from `SELECT column0 FROM table0 UNION SELECT column1 FROM table1 WHERE a = 1`,
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: s/a AST node/the AST

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+We know the AST is a type of tree structure of which child node corresponding to SQL text. 
+So we can walk through AST tree layer by layer and splice SQL text according to AST node. 
+There is a AST node parsed from `SELECT column0 FROM table0 UNION SELECT column1 FROM table1 WHERE a = 1`, 
+we can splice SQL text as the picture shows.
+
+![ast tree](./imgs/ast-tree.png)
+
+## Rationale
+
+See: [Rationale (Chinese)](https://docs.google.com/document/d/1KF4OvObvAzT0YW_KwiMRrsGZQMNVE03RdqShhStM_D4/edit?usp=sharing)
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: Should we translate a few conclusive sentences and put them here? @leoppro

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+## Compatibility
+
+None
+
+## Implementation
+
+### Code
+
+There is a `Recoverable` interface in [pingcap/parser/ast.go](https://github.com/pingcap/parser/blob/master/ast/ast.go).
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: The `Recoverable` interface is gone, replaced by a `Restore` method

Comment (tiancaiamao) [reply docs/design/2018-11-29-ast-to-sql-text.md]: Link a PR is better, the code changes always.

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+We need implement it for all `ast.Node()`.
+
+### Example
+
+I implemented the `Restore` function of `ast.CreateDatabasesStmt` and `ast.DropDatabaseStmt` for examples.
+
+```go
+// Restore implements Recoverable interface.
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: Stale, don't forget to update the comment in the following PRs

File: docs/design/2018-11-29-ast-to-sql-text.md (hunk 6 above)
Comment (XuHuaiyu) [inline docs/design/2018-11-29-ast-to-sql-text.md]: Do not link the Chinese doc here.
Please translate the rationale into English.

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+comment := Commentf("source %v \nrestore %v", t.src, restoreSQL)
+restoreStmt, err := parser.ParseOneStmt(restoreSQL, "", "")
+c.Assert(err, IsNil, comment)
+ast.CleanNodeText(stmt)
+ast.CleanNodeText(restoreStmt)
+c.Assert(restoreStmt, DeepEquals, stmt, comment)
+```
+
+We believe that a `Restore()` function is correct when it outputs sql text 
+can be parsed successfully and the AST correspond to output sql equals with input sql.
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: From the code, it's better to say:
when the SQL text it outputs can be parsed successfully and the generated AST equals with the original AST.

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+For other `ast.Node`:
+
+Some `ast.Node` can't be restored to a complete SQL text,
```
Comment (tiancaiamao) [inline docs/design/2018-11-29-ast-to-sql-text.md]: such as an expression

File: docs/design/2018-11-29-ast-to-sql-text.md (hunk 6 above)
Comment (leoppro) [reply docs/design/2018-11-29-ast-to-sql-text.md]: i will do it

Comment (leoppro) [reply docs/design/2018-11-29-ast-to-sql-text.md]: i remove it, i think the doc is not helpful to understanding this proposal

Comment (leoppro) [reply docs/design/2018-11-29-ast-to-sql-text.md]: i remove it, i think the doc is not helpful to understanding this proposal

**Comment by leoppro** [comment]
@tiancaiamao @XuHuaiyu PTAL again

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+splice SQL text according to AST node. 
+There is the AST parsed from `SELECT column0 FROM table0 UNION SELECT column1 FROM table1 WHERE a = 1`, 
+we can splice SQL text as the picture shows.
+
+![ast tree](./imgs/ast-tree.png)
+
+## Compatibility
+
+The AST and SQL text are one-to-many relationships, so we can't restore completely equal SQL text, 
+we only need to ensure that the ASTs parsed by input SQL and restored SQL are equal.
```
Comment (leoppro) [inline docs/design/2018-11-29-ast-to-sql-text.md]: parsed from

**Comment by tiancaiamao** [comment]
I'm fine with the content.
@CaitinChen would you like to take a look ?

**Comment by CaitinChen** [comment]
@tiancaiamao OK

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+So we can walk through the AST layer by layer, recursive `Restore()` of each child node and 
+splice SQL text according to AST node. 
+There is the AST parsed from `SELECT column0 FROM table0 UNION SELECT column1 FROM table1 WHERE a = 1`, 
+we can splice SQL text as the picture shows.
+
+![ast tree](./imgs/ast-tree.png)
+
+## Compatibility
+
+The AST and SQL text are one-to-many relationships, so we can't restore completely equal SQL text,
```
Comment (CaitinChen) [inline docs/design/2018-11-29-ast-to-sql-text.md]:
```suggestion
The AST and SQL text are in a one-to-many relationship, so we can't restore completely equal SQL text. 
```

File: docs/design/2018-11-29-ast-to-sql-text.md (hunk 11 above)
Comment (CaitinChen) [inline docs/design/2018-11-29-ast-to-sql-text.md]:
```suggestion
We only need to ensure that the ASTs parsed from the input SQL statement and the restored SQL statement are equal.
```

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+## Compatibility
+
+The AST and SQL text are one-to-many relationships, so we can't restore completely equal SQL text, 
+we only need to ensure that the ASTs parsed by input SQL and restored SQL are equal.
+
+## Implementation
+
+### Stage
+
+Considering that some ast.Node depend on another ast.Node, we divide sub-tasks into four stages.
```
Comment (CaitinChen) [inline docs/design/2018-11-29-ast-to-sql-text.md]:
```suggestion
Considering that some ast.Node depends on another ast.Node, we divide sub-tasks into four stages.  
```

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	sb.WriteString("DROP DATABASE ")
+	if n.IfExists {
+		sb.WriteString("IF EXISTS ")
+	}
+	WriteName(sb, n.Name)
+	return nil
+}
+```
+
+**There are another examples which is including complete implement and test:
```
Comment (CaitinChen) [inline docs/design/2018-11-29-ast-to-sql-text.md]:
```suggestion
**There are other examples which include complete implementation and test:
```

**Comment by CaitinChen** [comment]
And the title -> proposal: support restoring SQL text from an AST tree #8533

**Comment by leoppro** [comment]
LTAP again @tiancaiamao @XuHuaiyu

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**